### PR TITLE
Integration tests spec

### PR DESCRIPTION
**Pull request by mnovelodou**
## Summary

- Adds `integration-tests-spec.md` documenting the planned integration test suite
- Covers scope, technical setup, assumptions, and a full test inventory for `AccountController` and `ExpenseController`
- No code changes — spec only, ready for review and comments before implementation begins

## Test plan

- [ ] Review assumptions section and flag anything incorrect
- [ ] Review test inventory for missing or unnecessary cases
- [ ] Confirm open questions (setup via MockMvc vs direct JDBC, branch naming, etc.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: integration-tests-spec.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+---
+
+## Open questions
+
+These are not blocking — the spec records the decision made for each, but they should be confirmed before implementation.
+
+| Question | Decision taken |
+|---|---|
+| Should expense tests set up prerequisite accounts via MockMvc or direct JDBC injection? | Via MockMvc — keeps the full create path exercised and stays within the `@Transactional` boundary. |
```
Comment (mnovelodou) [inline integration-tests-spec.md]: Use MockMvc

File: integration-tests-spec.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+| Should we test the FK violation (delete account with expenses)? | Yes — it's a real edge case with a defined error shape (`500 DATABASE_ERROR`). |
```
Comment (mnovelodou) [inline integration-tests-spec.md]: Sure, lets add an integration tests like that, honestly this kind of test should error Http-412 Precondition Failed, API is NOT ready to handle that, so we need to implement it.

File: integration-tests-spec.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+| Should the base class be abstract or use a JUnit 5 extension? | Abstract base class — simpler, no extra dependency. |
```
Comment (mnovelodou) [inline integration-tests-spec.md]: Use Abstract base class

File: integration-tests-spec.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+| Test class naming: `*IT` or `*IntegrationTest`? | `*IT` — Maven/Gradle Failsafe convention; also clearly distinct from `*Test` unit tests. |
```
Comment (mnovelodou) [inline integration-tests-spec.md]: use *IT as per Maven convention

File: integration-tests-spec.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+| A14 | Not found | none (id 999999) | 404, `code: NOT_FOUND` |
+| A15 | Missing name | create account | 400, `code: BAD_REQUEST` |
+
+#### DELETE /accounts/{id}
+
+| # | Scenario | Setup | Expected |
+|---|---|---|---|
+| A16 | Happy path — deletes account | create account | 204, subsequent GET returns 404 |
+| A17 | Not found | none (id 999999) | 404, `code: NOT_FOUND` |
+| A18 | FK violation — account has expenses | create account + create expense against it | 500, `code: DATABASE_ERROR` |
```
Comment (mnovelodou) [inline integration-tests-spec.md]: Tecnically this should be HTTP-412

File: specs/integration-tests-spec.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
+# Integration Tests Spec
```
Comment (mnovelodou) [inline specs/integration-tests-spec.md:1]: Can you create an specs/ directory and move this file to specs/ ?

File: integration-tests-spec.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+---
+
+## Technical setup
+
+| Decision | Choice | Reason |
+|---|---|---|
+| Spring context | `@SpringBootTest(webEnvironment = MOCK)` | Full context + MockMvc. MOCK keeps the servlet call on the same thread as the test, making `@Transactional` rollback reliable. |
+| Database | Testcontainers (`org.testcontainers:postgresql`) | Real Postgres, isolated per test run, no shared state with local dev DB. |
+| Schema init | Reuse `classpath:/db/schema.sql` via `spring.sql.init.*` | Schema already uses `CREATE TABLE IF NOT EXISTS` — idempotent, no Flyway needed. |
```
Comment (mnovelodou) [inline integration-tests-spec.md]: This should explicitly require `spring.sql.init.mode=always` in the Testcontainers test properties. Boot's SQL initializer only runs automatically for embedded databases by default; with a PostgreSQL container, relying on `spring.sql.init.*` without forcing `mode=always` can leave the schema unapplied and make the first repository call fail with missing-table errors.

**Review by mnovelodou — COMMENTED** [review]
I left a few inline concerns where the spec needs tightening against the current app behavior.

File: specs/integration-tests-spec.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+```
+
+---
+
+## Assumptions
+
+1. **No cascade deletes.** `expenses.account_id` has a FK to `accounts` with no `ON DELETE CASCADE`. Attempting to delete an account that has expenses will produce a `DataIntegrityViolationException` → caught by `GlobalExceptionHandler.handleDatabase` → `500 DATABASE_ERROR`. This is tested explicitly.
+
+2. **`initialAmount` sets both columns on create.** `AccountRepository.create` inserts `initial_amount` and `current_amount` both from `entity.initialAmount()`. On update, only `current_amount` changes. Tests will assert this.
+
```
Comment (mnovelodou) [inline specs/integration-tests-spec.md:46]: I think the inventory should include a negative case for a missing/flat wrapper on accounts. `CreateAccountRequest` deserializes flat account JSON as `value == null`, and `AccountService.create` immediately calls `request.value()`/`validateAccountName(account)`, so this currently falls through to the generic 500 instead of a 400. Expense create already validates a missing payload, so this integration test would catch a real controller/service gap on the account path.

File: integration-tests-spec.md [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+---
+
+## Response shapes for reference
+
+**Error body (all 4xx/5xx):**
+```json
+{ "code": "NOT_FOUND", "message": "Account with ID 99 not found." }
+```
+
+**Account body:**
```
Comment (mnovelodou) [inline integration-tests-spec.md]: These reference shapes are unwrapped, but `POST /accounts`, `PUT /accounts/{id}`, and `POST /expenses` return `CreateAccountResponse`, `UpdateAccountResponse`, and `CreateExpenseResponse`, each with the domain object under `value`. The spec should call out which endpoints return the bare object (`GET /accounts/{id}`) versus `{ "value": { ... } }`, otherwise the future MockMvc assertions are likely to target `$.accountId`/`$.expenseId` when the actual paths are `$.value.accountId`/`$.value.expenseId` for write responses.
